### PR TITLE
Fusion jit

### PR DESCRIPTION
**Pull request by wtfnukee**
Hi! I've been playing with your framework and ended up writing a few things for the CPU backend as compiler project/takehome.

## Pointwise fusion JIT

A chain of elementwise operators is lowered to C, compiled by the host toolchain and loaded back with `mag_dylib`. The compiled kernel evaluates the whole chain per element, so intermediates stay in registers instead of being written out and read back once per operator.

There are two ways in. An explicit region:

```python
with no_grad(), mag.fuse():
    y = x * w + b
    z = y * y - x
```

and `optim.Adam`, which routes its ~20-operator update through one generated kernel automatically.

Inside a region the dispatcher records a fusible operator and marks its output pending instead of submitting it. Anything that cannot join the chain, and any read of a pending value, flushes it first. The guard sits in `mag_tensor_data_ptr`, which is the only way to reach a tensor's bytes.

Immediates are passed at call time, so changing a learning rate reuses the kernel. Kernels are cached in memory and on disk, keyed on the generated source: compilation costs 44 ms on the first call, 0.007 ms on later calls in the same process, and 0.25 ms in a new process reusing the cached object. Everything is built with `-ffp-contract=off`, and the plan mirrors the eager chain operator for operator, so fused and eager results are bit-identical. Only exact-IEEE operators are fusible; transcendentals end a chain, because the vector kernels approximate them and generated libm calls would not match. `MAG_JIT=off` disables the whole thing, and with no host compiler it falls back to eager.

A value is written out only if something outside the chain can still read it. Under gradient recording that requires knowing whether a backward reads an operand's data or only its shape, which `mag_op_backward_ignores_value` declares per operator.

## CPU kernels

The generic strided clone recomputed a full index decomposition per element and visited the source in destination order. It now walks the innermost two dimensions in tiles.

`masked_fill` walks runs instead of decomposing an index per element when the mask broadcasts over leading dimensions, which is the case for the causal mask in attention: `(1,1,T,T)` against `(B,H,T,T)`, where mask elements repeat.

The intra-op threading thresholds were about 100x too low, so small operators paid ~20 us of fan-out to save less than that. The crossover varies by 256x across operators. `benchmark/python/tune_intraop.py` re-derives the table and `MAG_CPU_INTRAOP_MIN_ELEMS` pins it.

`strided_view` backward now handles bijective views directly. Reshapes and permutations are bijections onto the base, so the gradient is a relabelling, not a scatter through an index tensor. Every `nn.Linear` weight gradient takes this path via `x @ weight.T`.

## Correctness

Unary operators computed different results depending on where an element landed. The vector kernels approximate `tanh`, `sigmoid`, `silu`, `sin`, `tan` and `gelu_approx`, while the scalar tail called libm. `tanh(0.7)` was `0.604367793` for length 3 and `0.604367733` for length 4, and a transposed view disagreed with a contiguous tensor. Every element now goes through the vector form.

The bfloat16 vector store truncated. The NEON store shifted right by 16 with no rounding bias while the scalar path rounded to nearest even, so `0.9 * 0.012` gave `0x3c31` at length 3 and `0x3c30` at length 4.

Float-to-integer casts were undefined behaviour. A plain `(TDst)x` meant the vectorized path saturated and the scalar path wrapped, so `-2.0` to `uint8` was 254 in a short tensor and 0 in a long one. There is one rule everywhere now: NaN to zero, saturate into the 32-bit range of the target's signedness, truncate to width.

RMSNorm could not be used in training, because `forward` used an in-place `sqrt_`, which is rejected whenever the input requires grad.

`examples/gpt2` could not run, and would not have been correct if it had: it imported a module removed in the nanobind migration, and wrote weights into `state_dict()`, which returns clones. Logits now match the reference to 0.035 across all twelve layers.

## Numbers

Autoencoder training step (`examples/ae` shape, median of 200, interleaved A/B):

| dtype | before | after | speedup |
|---|---|---|---|
| float32 | 2.66 ms | 0.87 ms | 3.0x |
| bfloat16 | 1.92 ms | 1.72 ms | 1.12x |

Roughly 2.2x of the float32 gain comes from the kernel and autograd work, the rest from the JIT.

Fusion, 8-operator float32 chain under `no_grad`:

| elements | eager | fused | speedup |
|---|---|---|---|
| 4 096 | 4.5 us | 2.5 us | 1.8x |
| 65 536 | 55.4 us | 12.5 us | 4.4x |
| 1 048 576 | 1015.6 us | 186.9 us | 5.4x |
| 4 194 304 | 5704.8 us | 953.2 us | 6.0x |

With gradient recording, on a 6-operator chain, the win depends on whether the backward needs the intermediates:

| chain | 64K | 1M |
|---|---|---|
| add (backward reads no operands) | 4.2x | 5.3x |
| mul (every intermediate is read) | 1.2x | 1.1x |

Before committing to code generation I compared it against interpreting the same chain, on 1M elements:

| approach | time | vs eager |
|---|---|---|
| eager, one op at a time | 464 us | 1.00x |
| tile interpreter | 398 us | 1.17x |
| register-machine interpreter | 551 us | 0.84x |
| generated code | 75 us | 6.17x |

Kernels:

| | before | after | speedup |
|---|---|---|---|
| transpose clone 512x1376 | 343 us | 73 us | 4.7x |
| transpose clone 1024x4096 | 3950 us | 1476 us | 2.7x |
| `masked_fill` 1x12x512x512 | 12.65 ms | 1.12 ms | 11.3x |
| strided `tanh` 512x512 | 423 us | 267 us | 1.6x |
| Adam step, 1M float16 params | 3.4 ms | 2.6 ms | 1.3x |

Transformer block step goes from 19.6 to 16.5 ms, and the run-to-run spread collapses from 19.6-57.0 ms to 16.5-16.7 ms, since the blocked walk is far less sensitive to cache contention.

GPT-2 124M, float32, KV cache:

| | before | after | speedup |
|---|---|---|---|
| prefill 128 tokens | 393.9 ms | 383.5 ms | 1.03x |
| prefill 512 tokens | 692.7 ms | 558.5 ms | 1.24x |
| prefill 1024 tokens | 1847.3 ms | 1332.1 ms | 1.39x |

Decode is 10.1 ms/token (98.8 tok/s), and 9.0 ms/token (110.0 tok/s) inside `mag.fuse()`. It reads all 124M weights per token, so it is bound by weight bandwidth rather than by anything fusion touches: matmul is 84% of the step.

### CODE REVIEW

**Comment by MarioSieg** [comment]
Thanks a lot for the contribution - there’s some really interesting work here, especially the fusion results and the CPU kernel improvements!

I’m not going to merge the JIT/fusion part in its current form, though. Magnetron needs a proper graph-level JIT/fusion architecture that is backend-agnostic, rather than introducing fusion specifically through the CPU backend. I’d like the graph representation and optimization/fusion passes to be shared across backends, with each backend responsible only for lowering/compiling that graph to its own implementation.

So while the generated-C approach is a cool proof of concept and the numbers are compelling, I don’t want to commit the project to a CPU-specific JIT architecture before that generic graph layer exists.

Some of the independent CPU fixes and optimizations in this PR look useful on their own, though, and I’d be happy to look at those separately if you want to split them out.

**Comment by wtfnukee** [comment]
Thanks for the guidance! I reworked the fusion part in #62 around a shared graph in core. Core now records the unfused dependencies and runs the liveness, grouping, boundary materialization, and scheduling passes. CPU and CUDA only lower and compile the selected graph through their own implementations.

I also split the independent CPU fixes and optimizations into #64.
